### PR TITLE
Dynamic bridge replay tests

### PR DESCRIPTION
**Pull request by VenelinMartinov**
This changes improves on the dynamically bridged provider TF GRPC recording and adds a test example of replaying the recordings in tests.

This also adds an example with pulumitests which is still waiting on a pu/pu release.

fixes https://github.com/pulumi/pulumi-terraform-provider/issues/37

### CODE REVIEW

File: dynamic/provider_test.go
Comment (guineveresaenger) [inline dynamic/provider_test.go:557]: What behavior does this test?

File: dynamic/provider_test.go
Comment (guineveresaenger) [inline dynamic/provider_test.go:599]: What does it mean to use the LogReplayProvider with a program? What is the expected behavior?

File: dynamic/internal/shim/grpcutil/translate.go
Comment (guineveresaenger) [inline dynamic/internal/shim/grpcutil/translate.go:117]: Can we add a comment that details the purpose of a LogReplayProvider?

**Review by guineveresaenger — COMMENTED** [review]
Can we provide a detailed pull request message, or code comment on `LogReplayProvider`, that outlines how this change implements the feature described in https://github.com/pulumi/pulumi-terraform-provider/issues/37? 
One thing that looks like it's missing to close https://github.com/pulumi/pulumi-terraform-provider/issues/37 is the presence and usage of `PULUMI_DEBUG_TF_LOGS` - unless that was added in a previous change, or there's a different mechanic we'll use.

File: dynamic/internal/shim/grpcutil/translate.go
Comment (iwahbe) [reply dynamic/internal/shim/grpcutil/translate.go]: It's generally not a great idea to distinguish between `[]T(nil)` and `[]T{}`. Since Go forces us to deal with `[]T(nil)` as the zero value of `[]T`, we try to stick to that for simplicity.

Another example is here:

https://github.com/pulumi/pulumi/blob/bd63383a89b6a6761d757af494b96ae12ee76e55/sdk/go/common/slice/slice.go#L17-L25

**Review by iwahbe — APPROVED** [review]
We're planning to tell users to use the replay recording feature. I think we need a test that setting `PULUMI_DEBUG_GRPC_LOGS=pulumi-logs.json PULUMI_TF_DEBUG_GRPC=tf-logs.json` will write the logs to a file, and that the logs are then usable to run a replay test. It looks like we have testing that given a log, we can replay with it, but no testing on getting a log written out.

File: dynamic/README.md
Comment (t0yv0) [inline dynamic/README.md:279]: For my 2c, this user experience - what we are asking users to do - is not great. Can we do better? Even if we ship this for now, what would it take to make this work seamlessly with

https://www.pulumi.com/docs/iac/support/troubleshooting/

File: dynamic/internal/shim/grpcutil/replay.go
Comment (t0yv0) [inline dynamic/internal/shim/grpcutil/replay.go:59]: The ordering of incoming requests will not be reliable in the general case. This will need a customizable matcher strategy from request to which log, and the default one probably should be a bit more aggressive than just matching on method + position. Not blocking on this but this will 100% bite us and need further elaboration.

File: dynamic/README.md
Comment (VenelinMartinov) [reply dynamic/README.md:279]: Yeah, I believe we discussed having the GRPC logs in the normal logs and then we can just grep the right lines in it. Opened https://github.com/pulumi/pulumi-terraform-bridge/issues/2535 for this.

File: dynamic/internal/shim/grpcutil/replay.go
Comment (VenelinMartinov) [reply dynamic/internal/shim/grpcutil/replay.go:59]: I haven't hit any issues with this yet and have opted for something simple here and we can extend later.

Comment (t0yv0) [reply dynamic/internal/shim/grpcutil/replay.go:60]: I'd like to push back on this. This is 100% programming in pain for maintainers. Can we at least have a tracking issue with cross-linked TODO in the code?

Comment (VenelinMartinov) [reply dynamic/internal/shim/grpcutil/replay.go:60]: Opened https://github.com/pulumi/pulumi-terraform-bridge/issues/2571

**Pull request merged**
